### PR TITLE
finding best chip region for pivot kick

### LINKED ISSUE #1878: Chip Ball Away More Intelligently
### Description of the task
In GoalieTactic, CreaseDefenderTactic, DefenseShadowEnemyTactic, and ShadowEnemyTactic, we define a fixed value (`YEET_CHIP_DISTANCE_METERS` in https://github.com/UBC-Thunderbots/Software/pull/1871) to yeet the ball away. With https://github.com/UBC-Thunderbots/Software/pull/1871, we can specify how far we chip the ball, so we should figure out the best place to chip the ball. For example, the default fixed value could chip the ball into the enemy's hands. 

### Acceptance criteria

- [ ] A more intelligent chip away system
- [ ] unit and/or simulated testing

### PR DESCRIPTION
**Pull request by xyztiger**
### Description

calculate best angle to chip within a rectangle approximation for pivot kick

### Resolved Issues

resolves #1878

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/evaluation/find_open_areas.h [hunk 1]
```diff
@@ -12,4 +13,5 @@
  * @return a vector of circles where the center is a good point to chip to, and the
```
Comment (jordanzliu) [inline src/software/ai/evaluation/find_open_areas.h:16]: u forgot to fix the javadoc comment

File: src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp [hunk 2]
```diff
@@ -210,15 +212,32 @@ void GoalieFSM::updatePivotKick(
         getNoChipRectangle(event.common.world.field()).xMax() + ROBOT_MAX_RADIUS_METERS;
     Point clear_origin = Point(clear_origin_x, event.common.world.ball().position().y());
 
-    Angle clear_direction = (event.common.world.ball().position() -
-                             event.common.world.field().friendlyGoalCenter())
-                                .orientation();
+    Point goalie_pos     = event.common.robot.position();
+    Point clear_corner_1 = Point(goalie_pos.x() + MAX_CHIP_DISTANCE - 1, 1.8);
```
Comment (jordanzliu) [inline src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp]: magic numbers should be named constants

**Review by jordanzliu — CHANGES_REQUESTED** [review]

File: src/software/ai/evaluation/find_open_areas.cpp [hunk 3]
```diff
@@ -1,26 +1,33 @@
...
 
-std::vector<Circle> findGoodChipTargets(const World& world)
+std::vector<Circle> findGoodChipTargets(const World& world,
+                                        std::optional<Rectangle> target_area_rectangle)
 {
     double inset     = 0.3;  // Determined experimentally to be a reasonable value
     double ballX     = world.ball().position().x();
     double fieldX    = world.field().enemyGoalCenter().x() - inset;
     double negFieldY = world.field().enemyCornerNeg().y() + inset;
     double posFieldY = world.field().enemyCornerPos().y() - inset;
```
Comment (itsarune) [inline src/software/ai/evaluation/find_open_areas.cpp]: this isn't related to your changes but by convention, we use ball_x, field_x, neg_field_y and pos_field_y

File: src/software/ai/evaluation/find_open_areas.cpp [hunk 4]
```diff
@@ -1,26 +1,33 @@
 #include "software/ai/evaluation/find_open_areas.h"
 
+#include <optional>
+
 #include "proto/parameters.pb.h"
 #include "software/geom/algorithms/find_open_circles.h"
 
-std::vector<Circle> findGoodChipTargets(const World& world)
+std::vector<Circle> findGoodChipTargets(const World& world,
```
Comment (itsarune) [inline src/software/ai/evaluation/find_open_areas.cpp:8]: could you add some validation for this?

**Review by itsarune — COMMENTED** [review]
looks good when I tested it locally, can you add some validation for it

File: src/software/ai/evaluation/find_open_areas.cpp (hunk 4 above)
Comment (xyztiger) [reply src/software/ai/evaluation/find_open_areas.cpp:8]: added unit tests

**Pull request closed without merging**